components/suggestion/emoticon_provider.jsx

```diff
@@ -49,82 +53,88 @@ class EmoticonSuggestion extends Suggestion {
-                    matched.push({name, emoji});
+        if (partialName.length < MIN_EMOTICON_LENGTH) {
+            SuggestionStore.clearSuggestions(suggestionId);
+            return false;
+        }
+
+        // Check for text emoticons if this isn't for an emoji reaction
+        if (prefix !== '-' && prefix !== '+') {
+            for (const emoticon of Object.keys(Emoticons.emoticonPatterns)) {
+                if (Emoticons.emoticonPatterns[emoticon].test(text)) {
+                    // Don't show the autocomplete for text emoticons
+                    return false;
                 }
             }
+        }
 
-            // Sort the emoticons so that emoticons starting with the entered text come first
-            matched.sort((a, b) => {
-                const aName = a.name;
-                const bName = b.name;
+        autocompleteCustomEmojis(partialName)(store.dispatch, store.getState).then(() => this.findAndSuggestEmojis(suggestionId, text, partialName));
 
-                const aPrefix = aName.startsWith(partialName);
-                const bPrefix = bName.startsWith(partialName);
+        this.findAndSuggestEmojis(suggestionId, text, partialName);
 
-                if (aPrefix === bPrefix) {
-                    return aName.localeCompare(bName);
-                } else if (aPrefix) {
-                    return -1;
-                }
-
-                return 1;
-            });
+        return true;
+    }
 
-            const terms = matched.map((item) => ':' + item.name + ':');
+    findAndSuggestEmojis(suggestionId, text, partialName) {
+        const matched = [];
 
-            SuggestionStore.clearSuggestions(suggestionId);
-            if (terms.length > 0) {
-                SuggestionStore.addSuggestions(suggestionId, terms, matched, EmoticonSuggestion, text);
+        const emojiMap = getEmojiMap(store.getState());
 
-                hasSuggestions = true;
+        // Check for named emoji
+        for (const [name, emoji] of emojiMap) {
+            if (emoji.aliases) {
+                // This is a system emoji so it may have multiple names
+                for (const alias of emoji.aliases) {
+                    if (!EMOJI_CATEGORY_SUGGESTION_BLACKLIST.includes(emoji.category) && alias.indexOf(partialName) !== -1) {
+                        matched.push({name: alias, emoji});
+                        break;
+                    }
+                }
+            } else if (name.indexOf(partialName) !== -1) {
+                // This is a custom emoji so it only has one name
+                matched.push({name, emoji});
             }
         }
 
-        if (hasSuggestions) {
-            // force the selection to be cleared since the order of elements may have changed
-            SuggestionStore.clearSelection(suggestionId);
+        // Sort the emoticons so that emoticons starting with the entered text come first
+        matched.sort((a, b) => {
```

Comment:
I feel like this method should be moved out of here, for easier reading.

    matched = sortWithPreferenceGivenToPrefixMatches(matched, partialName);